engine/src/main/java/org/pentaho/di/trans/steps/textfileoutput/TextFileOutput.java

```diff
@@ -0,0 +1,1076 @@
+          CompressionOutputStream compressionOutputStream = compressionProvider.createOutputStream( fileOutputStream );
+
+          // The compression output stream may also archive entries. For this we create the filename
+          // (with appropriate extension) and add it as an entry to the output stream. For providers
+          // that do not archive entries, they should use the default no-op implementation.
+          compressionOutputStream.addEntry( filename, environmentSubstitute( meta.getExtension() ) );
+
+          if ( log.isDetailed() ) {
+            if ( !Utils.isEmpty( meta.getEncoding() ) ) {
+              logDetailed( "Opening output stream in encoding: " + meta.getEncoding() );
+            } else {
+              logDetailed( "Opening output stream in default encoding" );
+            }
+          }
+
+          BufferedOutputStream bufferedOutputStream = new BufferedOutputStream( compressionOutputStream, 5000 );
+
+          fileStreams = data.new FileStream( fileOutputStream, compressionOutputStream, bufferedOutputStream );
+
+          data.getFileStreamsCollection().add( filename, fileStreams );
+
+          if ( log.isDetailed() ) {
+            logDetailed( "Opened new file with name [" + KettleVFS.getFriendlyURI( filename ) + "]" );
+          }
+        } else if ( fileStreams.getBufferedOutputStream() == null ) { // File was previously opened and now needs to be reopened.
+          int maxOpenFiles = getMaxOpenFiles();
+          if ( ( maxOpenFiles > 0 ) && ( data.getFileStreamsCollection().getNumOpenFiles() >= maxOpenFiles ) ) {
+            data.getFileStreamsCollection().closeOldestOpenFile( false );
+          }
+
+          OutputStream fileOutputStream = getOutputStream( filename, getTransMeta(), true );
+          CompressionProvider compressionProvider = getCompressionProvider();
+          CompressionOutputStream compressionOutputStream = compressionProvider.createOutputStream( fileOutputStream );
+          compressionOutputStream.addEntry( filename, environmentSubstitute( meta.getExtension() ) );
+          BufferedOutputStream bufferedOutputStream = new BufferedOutputStream( compressionOutputStream, 5000 );
+
+          fileStreams.setFileOutputStream( fileOutputStream );
+          fileStreams.setCompressedOutputStream( compressionOutputStream );
+          fileStreams.setBufferedOutputStream( bufferedOutputStream );
+        }
+      } catch ( Exception e ) {
+        if ( !( e instanceof KettleException ) ) {
+          throw new KettleException( "Error opening new file : " + e.toString() );
+        } else {
+          throw (KettleException) e;
+        }
+      }
+
+      fileStreams.setDirty( true );
+
+      data.fos = fileStreams.getFileOutputStream();
+      data.out = fileStreams.getCompressedOutputStream();
+      data.writer = fileStreams.getBufferedOutputStream();
+    } catch ( KettleException ke ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An exception is caught and re-thrown within the catch clause without any other operations being performed on the exception, making the catch clause unnecessary.

**Fix**: Perform additional operations on the exception that has been caught. You can wrap a standard exception into a custom exception to add additional information or special handling. When you do that, make sure to set the original exception as the cause. Otherwise, you will lose the stack trace and message of the original exception, which will make it difficult to analyze the event that caused the exception.  
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)